tests/firecrackerdrivertest.c

```diff
@@ -0,0 +1,302 @@
+        return -1;
+    }
+
+    VIR_DEBUG("xml read: %s", xml);
+
+    if (virDomainDefineXML(test_conn, xml) == NULL) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+testCreateDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+    int state = VIR_DOMAIN_NOSTATE;
+    int max_retries = 500;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainCreate(dom) < 0) {
+        return -1;
+    }
+
+    // Usually firecracker is set into running state even if the vm isn't
+    // yet ready to receive other commands like 'SendCtrlAltDel'
+    g_usleep(2500 * 1000);
+
+    do {
+        if (virDomainGetState(dom, &state, NULL, 0) < 0) {
+            VIR_DEBUG("Error getting state");
+            return -1;
+        }
+
+        if (max_retries <= 0)
+            return -1;
+
+        --max_retries;
+    }
+    while (state != VIR_DOMAIN_RUNNING);
+
+    return 0;
+}
+
+static int
+testShutdownDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+    int state = VIR_DOMAIN_NOSTATE;
+    int max_retries = 500;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainShutdown(dom) < 0) {
+        return -1;
+    }
+
+    do {
+        g_usleep(50 * 1000);
+
+        if (virDomainGetState(dom, &state, NULL, 0) < 0) {
+            VIR_DEBUG("Error getting state");
+            return -1;
+        }
+
+        if (max_retries <= 0)
+            return -1;
+
+        --max_retries;
+    }
+    while (state != VIR_DOMAIN_SHUTOFF);
+
+    return 0;
+}
+
+static int
+testUndefineDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainUndefine(dom) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+testCloseConnection(const void *opaque G_GNUC_UNUSED)
+{
+    if (virConnectClose(test_conn) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+# define MY_SUBTEST(TESTFUNC, arg) \
+    do { \
+        fprintf(stderr, #TESTFUNC "   ... "); \
+        if (TESTFUNC(opaque) < 0) { \
+            fprintf(stderr, "FAIL"); \
+            return -1; \
+        } else { \
+            fprintf(stderr, "OK"); \
+        } \
+        fprintf(stderr, "\n"); \
+    } while (0)
+
+static int
+testLifecycle(const void *opaque G_GNUC_UNUSED)
+{
+    fprintf(stderr, "\n");
+
+    MY_SUBTEST(testInitiateConnection, opaque);
+    MY_SUBTEST(testDefineDomain, opaque);
+    MY_SUBTEST(testCreateDomain, opaque);
+    MY_SUBTEST(testShutdownDomain, opaque);
+    MY_SUBTEST(testUndefineDomain, opaque);
+    MY_SUBTEST(testCloseConnection, opaque);
+
+    return 0;
+}
+
+static int
+testDefineShutdown(const void *opaque G_GNUC_UNUSED)
+{
+    fprintf(stderr, "\n");
```

Comment:
doesn't `virTestRun` already print newlines?

Reply:
Okay, but I see that the DO_SUBTEST macro already prints newlines. Ideally we shouldn't relying on the subtest functions for proper formatting